source-oracle/source_oracle/api.py

```diff
@@ -0,0 +1,332 @@
+        with conn.cursor() as c:
+            sql_columns = ','.join(["t." + f.alias for (k, f) in OracleColumn.model_fields.items() if f.alias != 'COL_IS_PK' and not f.exclude])
+
+            owners_comma = "'" + "','".join(owners) + "'"
+            query = f"""
+            SELECT {sql_columns}, NVL2(c.constraint_type, 1, 0) as COL_IS_PK FROM all_tab_columns t
+                LEFT JOIN (
+                        SELECT c.owner, c.table_name, c.constraint_type, ac.column_name FROM all_constraints c
+                            INNER JOIN all_cons_columns ac ON (
+                                c.constraint_name = ac.constraint_name
+                                AND c.table_name = ac.table_name
+                                AND c.owner = ac.owner
+                                AND c.constraint_type = 'P'
+                            )
+                        ) c
+                ON (t.owner = c.owner AND t.table_name = c.table_name AND t.column_name = c.column_name)
+                WHERE t.owner IN ({owners_comma})
+            """
+
+            columns = []
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                row = {k: v for (k, v) in row.items() if v is not None}
+                columns.append(OracleColumn(**row))
+
+            return columns
+
+# Backfills operate on ROWID, which is a unique identifier for each row based on its
+# physical location. This identifier is consistent in a single transaction but can change between
+# transactions. We use the ROWID to order all rows of a table and capture all rows up until the last ROWID
+# we have identified for this table at the time of starting the backfill. It is possible that some of the rows from
+# the middle of the range move to a place after the last ROWID, or to a smaller value than the ROWIDs we have processed
+# thus being missed by the backfill process, however since the incremental changes are captured based on transaction IDs, and not ROWIDs
+# the incremental changes task will capture those updated rows, so we will eventually have all the documents
+
+# In order to avoid races between backfill and incremental tasks, we implement a ping-pong between the two,
+# meaning that they don't really run concurrently ever, but rather they run one after another. We start by running
+# a backfill task (sync_counter = 0), backfilling CHECKPOINT_EVERY documents, and then passing the ball to the
+# incremental task (sync_counter = 1), which will fetch updates since the last seen transaction ID up until the latest transaction
+# of the database. The ball is then passed back to backfill (sync_counter = 2). This ensures
+# that we do not capture incremental updates to documents and emit them while a backfill task may be reading and emitting an older
+# version of the same row afterwards. Incremental updates always come after a backfill operation.
```

Comment:
See these comments for more information on how the backfill and incremental tasks work together to ensure consistency